File: backend/src/utils/cleanup.ts
import fs from "node:fs";
import path from "node:path";
import { logger } from "../utils/logger.js";

/**
 * 清理服务端安装后产生的安装文件和日志
 * @param instancePath 实例目录路径
 */
export async function cleanupInstallFiles(instancePath: string): Promise<void> {
  const filesToClean: string[] = [];

  // 1. 收集 forge-*-installer.jar 文件
  const entries = await fs.promises.readdir(instancePath, { withFileTypes: true });
  for (const entry of entries) {
    if (entry.isFile() && /^forge-.*-installer\.jar$/.test(entry.name)) {
      filesToClean.push(path.join(instancePath, entry.name));
    }
  }

  // 2. 收集 fabric-installer.jar 文件
  const fabricInstaller = path.join(instancePath, "fabric-installer.jar");
  if (fs.existsSync(fabricInstaller)) {
    filesToClean.push(fabricInstaller);
  }

  // 3. 收集 installer.log 文件
  const installerLog = path.join(instancePath, "installer.log");
  if (fs.existsSync(installerLog)) {
    filesToClean.push(installerLog);
  }

  // 删除文件
  for (const file of filesToClean) {
    try {
      fs.unlinkSync(file);
      logger.info(`已清理安装文件: ${path.basename(file)}`);
    } catch (error) {
      logger.warn(`清理文件失败 ${path.basename(file)}: ${error}`);
    }
  }
}

File: core/util/cleanup.go
package util

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// CleanupInstallFiles removes installer artifacts after server installation
func CleanupInstallFiles(instancePath string) {
	filesToClean := []string{}

	// 1. Collect forge-*-installer.jar and forge-*-installer.jar.log files
	entries, err := os.ReadDir(instancePath)
	if err != nil {
		Logger.Warn(fmt.Sprintf("Failed to read instance directory: %v", err))
		return
	}

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}

		name := entry.Name()
		// Match forge installer files
		if strings.HasPrefix(name, "forge-") &&
			(strings.HasSuffix(name, "-installer.jar") ||
				strings.HasSuffix(name, "-installer.jar.log")) {
			filesToClean = append(filesToClean, filepath.Join(instancePath, name))
		}
	}

	// 2. Collect fabric-installer.jar file
	fabricInstaller := filepath.Join(instancePath, "fabric-installer.jar")
	if _, err := os.Stat(fabricInstaller); err == nil {
		filesToClean = append(filesToClean, fabricInstaller)
	}

	// 3. Collect installer.log file
	installerLog := filepath.Join(instancePath, "installer.log")
	if _, err := os.Stat(installerLog); err == nil {
		filesToClean = append(filesToClean, installerLog)
	}

	// Delete files
	for _, file := range filesToClean {
		err := os.Remove(file)
		if err != nil {
			Logger.Warn(fmt.Sprintf("Failed to clean file %s: %v", filepath.Base(file), err))
		} else {
			Logger.Info(fmt.Sprintf("Cleaned installer file: %s", filepath.Base(file)))
		}
	}
}